src/main/resources/META-INF/rewrite/jakarta-ee-9.yml

```
@@ -988,3 +990,15 @@ recipeList:
       groupId: jakarta.platform
       artifactId: "*"
       newVersion: 9.0.0
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.migrate.jakarta.RemoveJakartaAnnotationDependency
+displayName: Remove jakarta.annotation-api dependency if already manage by other dependency
+description: This recipe mainly revert the jakarta.annotation-api added by org.openrewrite.java.migrate.javax.AddCommonAnnotationsDependencies
+recipeList:
+  - org.openrewrite.java.dependencies.DependencyInsight:
+      groupIdPattern: jakarta.annotation
+      artifactIdPattern: jakarta.annotation-api
```

[Preconditions](https://docs.openrewrite.org/reference/yaml-format-reference#preconditions) are the typical way to limit what files recipes operate on. That would look something like this
```suggestion
preconditions:
  - org.openrewrite.java.dependencies.DependencyInsight:
      groupIdPattern: jakarta.annotation
      artifactIdPattern: jakarta.annotation-api
recipeList:
```

I'm not sure about the use of DependencyInsight here though, as we likely only want to remove the dependency if it's already available transitively. That might need some custom logic. 🤔

---

Not right now; we do have an `onlyDirect` option on that recipe, but perhaps we'd need to add an `onlyTransitive`:
https://docs.openrewrite.org/recipes/maven/search/dependencyinsight#dependencies-in-use

---

Possibly yes. Although that might not be too difficult as we already know the depth there, since we add that to the data table we produce.